### ISSUE TITLE
Dashboard for review comments

### ISSUE THREAD

**Issue by cmutel** [cauldron/bw_matchbox#93]
55c9e6e386eef439bec8b4f35dc7d56ba7089a3f introduced a data store and API endpoint for review comments. We now need a dashboard to display, sort, filter, and interact with these comments. I guess this would be similar to the `index` view, with a template that loads data from the API dynamically.

Here are the user stories that this dashboard should fulfill:

* As a reviewer, I want to see the status of my comments
* As a reviewer, I want to be able to see all comments which haven't been resolved
* As a reviewer, I want to be able to close or continue conversations from the dashboard view
* As an editor, I want to be able to see all comments
* As an editor, I want to be able to see all comments which haven't been resolved
* As an editor, I want to respond to comments from the dashboard view

To get test data, you can call the development endpoint `/generate-fake-comments` several times. There is also `/delete-all-comments` if that is helpful.

I leave it to @lilliputten to decide on the best way to display this threaded data, there are many options. My only request is that we have a reasonable agreement with the visual style of the rest of the app.

**Comment by lilliputten** [comment]
Are we going to get a user role tag (editor, reviewer) from the server (in the cookies)?

Should this page be accessible only to these two roles (and closed to all other users)?

Should the reviewer see (and control?) only his comments or all the comments of all users? (M.b. all the his own comments and all unresolved ones?)

Is each comment the beginning of a discussion thread (or chat)?

Probably it would be nice to have an access rights table for user roles (editor, reviewer)?

**Comment by cmutel** [comment]
> Are we going to get a user role tag (editor, reviewer) from the server (in the cookies)?

It is `{{ config.role }}` in the template.

> Should this page be accessible only to these two roles (and closed to all other users)?

Accessible to everyone. But *only the initial thread creator* can close a comment thread. I will add the username of the initial reporter to the comment API (#106)

> Should the reviewer see (and control?) only his comments or all the comments of all users? (M.b. all the his own comments and all unresolved ones?)

All comments are public (for people who can login at all). But only editors and reviewers can create comment threads and make comments (see user stories), and only initial reporters can resolve comment threads.

> Is each comment the beginning of a discussion thread (or chat)?

Yes, each comment is required to be in a thread, so the initial modal will need a "title" or similar attribute in addition to the comment content field.

> Probably it would be nice to have an access rights table for user roles (editor, reviewer)?

I want to avoid this is possible, making everything public is easy and there is little benefit to adding more complexity for this application.

**Comment by lilliputten** [comment]
For now the task seems to be incomplete. I'll try to ask questions to clarify the problem statement.

What interface elements we need to have on the page? (Filters, selectors, search, etc? Probable, the 'resolved' selector, at least?)

Do I need to have (and use?) that 'role' (for the current user) on the client side or is it only for the server side? I think, I'll need it at least for display proper operations on the particular comments and threads? How can I access this value on the client side?

Am I right assuming that when I request `/comments-api` without any parameters I get all the available comments?

How can I get the list of thread's roots (only root comments)? (See my idea of `position` field usage below. Is it true?)

How can I determine 'root' and 'regular' comments? I can't see no parameters like `parent` or `children` (no linking data in the comment data model at all). Perhaps the hypothesis is true that the comment with minimal (or exact zero?) `position` property is the root of the thread?

Probably, I don't need to load only roots first and only then (by demain) to load required threads? But instead load all the data (with empty request) and process it on the client?

Should I display a list of 'root comments' (starting comments for threads) by default?

How will we display the contents of the threads? By expand the rows? By open modals with thread contents?

What operations should we have for thread roots (threads themselves) and for each of paricular comments?

I suppose that possible options are: 'close' (is it the same as 'resolve'?), 'comment' (respond). What else ('delete', etc)?

Brief summary:

- Interface (required controls, filters and user operations), list of threads, way to get threads' content.
- Other server data (`role`?).
- Determining types of comment records and getting required data (threads' list, thread's content).

**Comment by cmutel** [comment]
> What interface elements we need to have on the page? (Filters, selectors, search, etc? Probable, the 'resolved' selector, at least?)

I think the easiest way to do this is a single page with filters at the top, like the way we present processes. The filters can be:

* Resolved state (no filter, resolved, open)
* Process (Can be a `select` element with `multiple`)
* Author (Again a `select` element)
* Participants (

Plus buttons to reset all filters, and "My comment threads", which filters to threads where my username is a participant, and are not resolved.

For now no search is needed.

## Permissions:

* In each comment thread, everyone can see the whole comment thread.
* The user who created the thread (i.e. `{{ config.user }}` is the same as `thread_reporter`) has a button to resolve the comment thread.
* Users with the role `editors` and `reviewers` have the option to add a new comment at the bottom of each comment thread, with a submit button.

> What other server data would be useful to have on the client?

After we build the dashboard with test data, we can add commenting to the process views, and then we can decide if we are missing some helpful functionality. 

> Am I right assuming that when I request /comments-api without any parameters I get all the available comments?

Yes

> How can I get the list of thread's roots (only root comments)? (See my idea of position field usage below. Is it true?)

Not possible, and not planned. 

> How can I determine 'root' and 'regular' comments? I can't see no parameters like parent or children (no linking data in the comment data model at all). Perhaps the hypothesis is true that the comment with minimal (or exact zero?) position property is the root of the thread?

There are no `root` comments, just a list of comments per thread with order given by `position` counting up from 0. As such, there are no parent or child comments. The first comment in a thread *should* always have position 0, but this isn't enforced. The comments API will always return in ascending order thread id and then by position.

> Probably, I don't need to load only roots first and only then (by demain) to load required threads? But instead load all the data (with empty request) and process it on the client?

Yes, I think so.

> Should I display a list of 'root comments' (starting comments for threads) by default?

This seems reasonable, but you could also imagine also expanding any threads the author is a participant in. We can also add `limit` and `offset`, but maybe it is OK to just have everything on one long page for now.

> How will we display the contents of the threads? By expand the rows? By open modals with thread contents?

I think just normal text elements using CSS to indent the comments in each thread. It would be nice to also have comments rendered through a markdown parser, this can happen server side if that is more sensible. We don't need fancy things like code rendering, just basic elements.

> What operations should we have for thread roots (threads themselves) and for each of paricular comments? 

Nothing more than the specifics listed above for certain users, and maybe the ability to collapse or expand a thread.

> I suppose that possible options are: 'close' (is it the same as 'resolve'?), 'comment' (respond). What else ('delete', etc)?

No delete, resolve instead of close (the thread will still be visible), and comment.

> Do we need some some kind of incremental (paginated) loading there (like on processes list page)?

No, not now.

**Comment by lilliputten** [comment]
Is it good to stroe threads data (`thread_name` and `thread_reporter`) in each comment object?

M.b. it would be better to have dedicated threads objects list with their own properties?

**Comment by cmutel** [comment]
> Is it good to store threads data (thread_name and thread_reporter) in each comment object?
> M.b. it would be better to have dedicated threads objects list with their own properties?

OK, can you give your preferred alternative to [what we have here](https://github.com/cauldron/bw_matchbox/blob/3d4947b31b9adfd2b7b54b8e19125cd87446c08e/bw_matchbox/webapp.py#L214)?

> Probalbly (just my thought) also it would be useful to sort threads by last activity time (by last comment creation time). To do it we need to have timestamp field.

Agreed, #108.

> What is thread's reporter? Should I show it close to the thread name, like it's the thread's author?

I was thinking of this only for you internally to decide whether to display the "resolve" action, but we can definitely include it next to the name of the comment thread.

**Comment by lilliputten** [comment]
> > Is it good to store threads data (thread_name and thread_reporter) in each comment object?
> > M.b. it would be better to have dedicated threads objects list with their own properties?
> 
> OK, can you give your preferred alternative to [what we have here](https://github.com/cauldron/bw_matchbox/blob/3d4947b31b9adfd2b7b54b8e19125cd87446c08e/bw_matchbox/webapp.py#L214)?

For example, this way (in typescript notation):

```typescript
interface TCommentProcess {
interface TCommentProcess {
  database: string; // 'new1'
  id: number; // 5300
  location: string; // 'United States'
  name: string; // 'Proxy for Ethalfluralin'
  product?: string; // null
  unit: string; // 'USD'
  url: string; // '/process/5300'
}
interface TComment {
  content: string; // '...'
  position: number; // 10
  resolved: boolean; // true
  thread_id: number; // <- `TThread['id']
  /* // UNUSED: Could be fetched from threads data
   * thread_name: string; // 'Consequatur exercita'
   * thread_reporter: string; // 'Reporter Name'
   */
  user: string; // 'Ida Trombetta'
  process: TCommentProcess;
}
interface TThread {
  id: number; // <- `TComment['thread_id']
  name: string;
  reporter: string;
}
interface TResponse {
  comments: TComment[];
  threads: TThread[];
  total_comments: number;
  total_threads: number;
}
```

IDK, is it neccessary/good or not, but here is possible to use ids for comments and list of these ids in each thread and drop `thread_id` property from comments' data, does it make sense?

**Comment by cmutel** [comment]
In ac193847525c3f4a7cac635698cd14d17e334e4a, I:

* Updated comment DB to include created and modified timestamps for the *comment thread*
* Changed comment API endpoint spec based on input from @lilliputten to:

```python
{
    "total_threads": int,
    "total_comments": int,
    "comments": [
        {
            "id": int, 1-indexed (from db),
            "thread": int, 1-indexed (from db),
            "content": str,
            "position": int, 0-indexed,
            "user": str,
        }
    ],
    "threads": [
        {
            "id": int, 1-indexed,
            "name": str,
            "reporter": str,
            "resolved": bool,
            "created": ISO 8601 str,
            "modified": ISO 8601 str,
            "process": {
                "id": int, 1-indexed (from db),
                "database": str,
                "name": str,
                "unit": str,
                "location": str,
                "product": str,
                "url": str, relative URL,
            },
        }
    ]
}
```

> IDK, is it neccessary/good or not, but here is possible to use ids for comments and list of these ids in each thread and drop thread_id property from comments' data, does it make sense?

If this is more convenient, it can easily be done in JS to the response data.

**Comment by lilliputten** [comment]
> * Resolved state (no filter, resolved, open)
> * Process (Can be a `select` element with `multiple`)
> * Author (Again a `select` element)
> * Participants (
> 
> Plus buttons to reset all filters, and "My comment threads", which filters to threads where my username is a participant, and are not resolved.

I started to implement the filters and have a few questions:

- What is the difference between `Author` and `Participants`?

Which of them should filter comments by author (by `user` field)? Should I use multiple choice (to filter for several users)?

What should do the other? Filter entire threads by `reporter`?

For 'Process', 'Author', 'Participants` filters do I have to populate filter lists from the actual data?

Should 'My comment threads' disable (or just reset) all the other filters?

**Comment by cmutel** [comment]
> What is the difference between Author and Participants?

Nothing, sorry, I would use "Comment author(s)" to make it clear. This is the filter for the comment author (`user` field).

> What should do the other? Filter entire threads by reporter?

Yes

> For 'Process', 'Author', 'Participants` filters do I have to populate filter options from the actual data?

Yes

> Should 'My comment threads' disable (or just reset) all the other filters?

Yes, reset all other filters

**Comment by lilliputten** [comment]
Trying to send `resolve-thread` request. Getting internal error. Bellow are detailed request and response infos.

Server method `comments_resolve_thread` doesn't process all the error types (only `DoesNotExist`) and returns standard error traceback in html. Q: Is it possible to return standartizied json error if smth happened in response to `application/json` requests?

General request:
```
Request URL: http://localhost:5000/comments/resolve-thread
Request Method: POST
Status Code: 500 INTERNAL SERVER ERROR
Remote Address: 127.0.0.1:5000
Referrer Policy: strict-origin-when-cross-origin
```

Request:
```
POST /comments/resolve-thread HTTP/1.1
Accept: application/json
Accept-Encoding: gzip, deflate, br, zstd
Accept-Language: en-US,en;q=0.9
Authorization: Basic am9obm55Om1uZW1vbmlj
Connection: keep-alive
Content-Length: 28
Content-Type: application/json
Cookie: project=matchbox-example; source=USEEIO-1.1; target=USEEIO-copy; proxy=new1; config=my_config
Host: localhost:5000
Origin: http://localhost:5000
Referer: http://localhost:5000/comments
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-origin
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/118.0.0.0 Safari/537.36
sec-ch-ua: "Chromium";v="118", "Google Chrome";v="118", "Not=A?Brand";v="99"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "Windows"
```

Payload:
```json
{"thread":3,"resolved":true}
```

Response:
```
HTTP/1.1 500 INTERNAL SERVER ERROR
Server: Werkzeug/2.3.6 Python/3.11.3
Date: Sun, 13 Aug 2023 18:36:12 GMT
Content-Type: text/html; charset=utf-8
Content-Length: 16806
Connection: close
```

Flask traceback:
```
127.0.0.1 - - [14/Aug/2023 01:36:12] "POST /comments/resolve-thread HTTP/1.1" 500 -
Traceback (most recent call last):
  File "bw_matchbox/.venv/Lib/site-packages/flask/app.py", line 2213, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "bw_matchbox/.venv/Lib/site-packages/flask/app.py", line 2193, in wsgi_app
    response = self.handle_exception(e)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "bw_matchbox/.venv/Lib/site-packages/flask/app.py", line 2190, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "bw_matchbox/.venv/Lib/site-packages/flask/app.py", line 1486, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "bw_matchbox/.venv/Lib/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "bw_matchbox/.venv/Lib/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "bw_matchbox/.venv/Lib/site-packages/flask_httpauth.py", line 174, in decorated
    return self.ensure_sync(f)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:/Work/upwork/230726-chris-mutel/bw_matchbox/bw_matchbox/webapp.py", line 216, in comments_resolve_thread
    thread["resolved"] = content["resolved"]
TypeError: 'CommentThread' object does not support item assignment
```

I'm stopping now and will contibue tomorrow. Good night!

**Comment by lilliputten** [comment]
Fixed in PR #109.

**Comment by lilliputten** [comment]
Small fixes in PR #110.
